### ISSUE TITLE
Closure inference fails to infer that `move` is required when pattern matching

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust#30046]
See [this example](http://is.gd/1atRZn), which fails to compile:

``` rust
fn foo<F>(f: F)
    where F: FnOnce()
{
}

fn main() {
    let x = Some(vec![1, 2, 3]);
    foo(|| {
        match x {
            Some(y) => { }
            None => { }
        }
    });
}
```

Error:

```
<anon>:9:15: 9:16 error: cannot move out of captured outer variable in an `FnOnce` closure
<anon>:9         match x {
                       ^
<anon>:10:18: 10:19 note: attempting to move value to here
<anon>:10             Some(y) => { }
                           ^
<anon>:10:18: 10:19 help: to prevent the move, use `ref y` or `ref mut y` to capture value by reference
error: aborting due to previous error
playpen: application terminated with error code 101
```

**Comment by nikomatsakis** [comment]
triage: P-medium

**Comment by durka** [comment]
A-const-fn?

**Comment by durka** [comment]
Do you just mean the help message is wrong, or it should have automatically created a move closure?

**Comment by nikomatsakis** [comment]
@durka neither. Normally, even for non-move closures, we infer when particular variables are used in the body and we move just those variables. For some reason, that is not triggering in this particular case.

**Comment by nagisa** [comment]
Might be duplicate of https://github.com/rust-lang/rust/issues/26139?

**Comment by nikomatsakis** [comment]
Related example from #26139: 

```rust
struct Name(String);

fn apply<F>(f: F) where
    F: FnOnce() {

    f()
}

fn main() {
    let n = Name(String::new());
    let consume = || {
        let Name(i) = n;
    };

    apply(consume);
}
```
